### ISSUE TITLE
Server SslStream Inner Exception -"The Local Security Authority cannot be contacted" when used with CNGKey with PIN

### ISSUE THREAD

**Issue by teenccu** [dotnet/runtime#36470]
I am creating CNGKey with TPM and creating a self-signed certificate. When I am using the certificate with SSLStream I am getting error "A call to SSPI failed, see inner exception." with inner exception  "The Local Security Authority cannot be contacted".

The Code for creating the certificate is as below.
_//Create the parameters
  CngKeyCreationParameters creationParameters = new CngKeyCreationParameters()
            {
               KeyCreationOptions = CngKeyCreationOptions.OverwriteExistingKey,
                KeyUsage = CngKeyUsages.AllUsages,
            Provider = new CngProvider("Microsoft Platform Crypto Provider"),
                ExportPolicy = CngExportPolicies.AllowArchiving
            };
//Setting the PIN
 const string NCRYPT_PIN_PROPERTY = "SmartCardPin";
CngProperty pinProperty = new CngProperty(
		        NCRYPT_PIN_PROPERTY,
		        propertyBytes,
		        CngPropertyOptions.None);
creationParameters.Parameters.Add(pinProperty);
//Create the Key
 CngKey key = CngKey.Create(CngAlgorithm.Rsa, "Test", creationParameters);
//Create Certficate
 using (RSA rsa = new RSACng(key))
 {
var request = new CertificateRequest(distinguishedName, rsa, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);
var usages = X509KeyUsageFlags.DataEncipherment | X509KeyUsageFlags.KeyEncipherment | X509KeyUsageFlags.DigitalSignature;
 if (IsCertificateAuthority) usages = usages | X509KeyUsageFlags.KeyCertSign;
request.CertificateExtensions.Add(new X509KeyUsageExtension(usages, false));
request.CertificateExtensions.Add(
                     new X509EnhancedKeyUsageExtension(
                             new OidCollection { new Oid("1.3.6.1.5.5.7.3.1") }, false));

request.CertificateExtensions.Add(sanBuilder.Build());
if (IsCertificateAuthority)
                    request.CertificateExtensions.Add(new X509BasicConstraintsExtension(true, true, 1, true));

if (ExpirationAfter == null) 
{ ExpirationAfter = DateTime.UtcNow.AddDays(-1).AddYears(10); }
if (ExpirationBefore == null) ExpirationBefore = DateTime.UtcNow;
}
 var certificate = request.CreateSelfSigned(new DateTimeOffset(ExpirationBefore.Value), new DateTimeOffset(ExpirationAfter.Value));
}_

However, if the PIN is not set this is working fine. 
Can anybody help me on the problem

**Comment by bartonjs** [comment]
> However, if the PIN is not set this is working fine.

I'm not sure what state you're describing.  Is it that if you make a key that doesn't use a PIN at all that everything just works? If so, then it's probably a limitation that SslStream/SChannel doesn't have a way of PIN prompting when they reopen the key in a different process (TLS is done out of process on Windows).

I'm not sure that Windows has a scenario where a TLS server cert is PIN protected, it may be an OS limitation.

**Comment by teenccu** [comment]
Yes, if I create a Key without a PIN , it is working correctly with SSLStream.

Probably it is a limitation with SSLStream  as it is asking for the PIN if I create key with PIN and then certificate in the same way and use it in WCF (WebHttpBinding) , it is asking for the PIN.

So I have two questions
1. Even for WCF , I want to put the PIN inside program and dont want the Popup. Is there any way to set the PIN in certificate handle and void the popup ?
2. Is there any workaround for SSLStream to make the certificate with privatekey having a PIN to work ?

**Comment by teenccu** [comment]
@bartonjs 

Thanks for your response. The Key prompt was for client certificate in case of mutual authentication.

I did some research and as you mentioned I saw that SSLStream uses the external process Isass.exe for opening the Key and there it fails if there is a PIN.

And in WCF, though the Certificate is attached to the channel is with a  CNGKey(with PIN property set) but later the Key  is again opened by the framework using a new Ncrypt Storage Provider and hence it asks the PIN again.

Finally, I solved the situation by writing a my own Key Storage Provider where I managed the situation of the key protection myself. I used the TSS library from Microsoft to access the TPM inside my Storage Provider. It worked with both SSLStream and WCF. It helped me to learn a bit deeper in the subject and I thank you all for the support.
